Refactor

LemonTea387 opened the pull request.

LemonTea387 committed:
update to use new cmake syntax

LemonTea387 committed:
update to glad2 and embed generated now instead of using cmake to gen.

LemonTea387 committed:
clangd file for std::expected.

LemonTea387 committed:
bump cxx standard to 23

LemonTea387 committed:
simple refactor for a reusable engine.

LemonTea387 committed:
Quick main.

LemonTea387 committed:
very fast setup system for writing and loading programs

Includes an example Triangle to demo.

LemonTea387 committed:
rename to match

LemonTea387 committed:
some proper program selection.

I cooked.

LemonTea387 committed:
add in some misc shader uniform stuff

LemonTea387 committed:
eh maybe it can be part of the main ctor

LemonTea387 committed:
ignore the imgui stuff

LemonTea387 committed:
add some todo tracking for readme.

LemonTea387 committed:
Fixup syntax typos for checkbox

LemonTea387 committed:
add more to plan

LemonTea387 committed:
slight update on ui

Diesel-Hadez commented on .clangd:
```diff
@@ -0,0 +1,2 @@
+CompileFlags:
+  Add: [-D__cpp_concepts=202211L]
```
With clang 19, this shouldn't be an issue anymore.  See: https://github.com/clangd/clangd/issues/2200#issuecomment-2451061144

Diesel-Hadez replied:
Why did you resolve this, I didn't see a commit removing this >:(

Diesel-Hadez commented on .gitignore:
```diff
@@ -1,3 +1,5 @@
 .cache
 compile_commands.json
+
+*.ini
```
What .ini files ??? For imgui? Can probably just make your setup run it from the build directory.

LemonTea387 replied:
Yea tbh

Diesel-Hadez replied:
Why did you resolve this, I didn't see a commit removing this >:(

Diesel-Hadez commented on .gitmodules:
```diff
@@ -0,0 +1,4 @@
+[submodule "lib/imgui"]
+	path = lib/imgui
+	url = https://github.com/ocornut/imgui.git
+	branch = docking
```
Consider instead of using a git submodule, using a package manager like vcpkg or conan (Or at the very least, CMake's FetchContent.)

Diesel-Hadez commented on CMakeLists.txt:
```diff
@@ -33,35 +20,59 @@ FetchContent_Declare(
+  # GLAD
+  "${CMAKE_SOURCE_DIR}/lib/glad/include"
+)
+
```
why not fetchcontent these too?

Diesel-Hadez commented on CMakeLists.txt:
```diff
@@ -33,35 +20,59 @@ FetchContent_Declare(
+
+add_executable(${PROJECT_NAME} 
+        ${DEPS}
+        ${SOURCE_FILES})
```
Try avoid glob recurse imo

Diesel-Hadez commented on CMakeLists.txt:
```diff
@@ -33,35 +20,59 @@ FetchContent_Declare(
-    add_subdirectory(${glfw_SOURCE_DIR} ${glfw_BINARY_DIR})
-endif()
+FetchContent_MakeAvailable(glfw glm)
 
```
Okay for now, but consider splitting into separate CMake functions in a separate file if it gets too large.

Diesel-Hadez commented on README.md:
```diff
@@ -1,2 +1,11 @@
-# Opengl-sandbox
-A place where I keep track of things to play around with opengl. 
+# Graphics-sandbox
```
Make it pretty, add [badges/shields](https://shields.io/), images of the project are important to showcase it, and maybe a CI/CD pipeline for automated deployment of the program (AppImage/exe/whatever), do formatting (clang-tidy/clang-format) and testing with unti tests before merging to main.

Diesel-Hadez commented on README.md:
```diff
@@ -1,2 +1,11 @@
+- [ ] Graphics backend abstraction, scenes can use a common interface.
+- [ ] Vulkan?
+- [ ] Better errors, currently it's just a raw enum that tells us 0 details.
+- [ ] An actual logging system, currently doing cerr and it's horrible.
```
Perhaps look into spdlog.

Alternatively, if you want to experiment with this, there is a way to configure cmake to autogenerate files and fill it with content. For example, you can autoconfigure it to generate a `SceneX__logging.hpp` for every `SceneX.hpp` you make, which makes a logging class inside it, useful because it automatically knows what class it is that is being logged.

Diesel-Hadez commented on lib/glad/README.md:
```diff
@@ -0,0 +1,8 @@
+# GLAD 4.1
+## Specs
+GLAD v2
+gl 4.1 Core
```
Formatting (e.g: Specifications. )

Diesel-Hadez commented on lib/glad/include/KHR/khrplatform.h:
```diff
@@ -0,0 +1,311 @@
+#ifndef __khrplatform_h_
```
Again, use a package manager of some sort.

Diesel-Hadez commented on src/engine/application.hpp:
```diff
@@ -0,0 +1,30 @@
+#include <cstdint>
+#include <string>
+
+#include "engine/window.hpp"
```
Unneeded. Just forward declare it.

Diesel-Hadez replied:
I saw your comment that your LSP did it. Fix your LSP to make it forward-declare it instead.

Diesel-Hadez commented on src/engine/application.hpp:
```diff
@@ -0,0 +1,30 @@
+
+struct ApplicationSpec {
+  std::string Title;
+  uint32_t Width;
```
Consider using `std::uint32` instead

Diesel-Hadez commented on src/engine/renderer/shader.cpp:
```diff
@@ -0,0 +1,130 @@
+void Shader::setFloat4(const std::string &name, const glm::vec4 &value) {
+  glUniform4f(glGetUniformLocation(m_Program, name.c_str()), value.x, value.y,
+              value.z, value.a);
+}
```
Maybe use `std::string_view` ? Not sure if it works.

Diesel-Hadez commented on src/engine/renderer/shader.cpp:
```diff
@@ -0,0 +1,130 @@
+  glUniform4f(glGetUniformLocation(m_Program, name.c_str()), value.x, value.y,
+              value.z, value.a);
+}
+void Shader::setMat4(const std::string &name, const glm::mat4 &value) {}
```
Use some error or logging or something to mark that this is not yet implemented and it is a TODO.

Diesel-Hadez commented on src/engine/renderer/shader.hpp:
```diff
@@ -0,0 +1,48 @@
+
+  const std::string &getName() const;
+
+  static std::expected<std::unique_ptr<Shader>, EngineError> create(
```
Consider `[[nodiscard]]` to make sure the return is always handled.

Diesel-Hadez commented on src/engine/renderer/shader.hpp:
```diff
@@ -0,0 +1,48 @@
+  void setFloat2(const std::string &name, const glm::vec2 &value);
+  void setFloat3(const std::string &name, const glm::vec3 &value);
+  void setFloat4(const std::string &name, const glm::vec4 &value);
+  void setMat4(const std::string &name, const glm::mat4 &value);
```
const qualifier

Diesel-Hadez commented on src/engine/window.cpp:
```diff
@@ -0,0 +1,33 @@
+
+void framebufSizeCallback(GLFWwindow* window, int width, int height) {
+  glViewport(0, 0, width, height);
+}
```
Either wrap this in an anonymous namespace, or replace this usage with a lambda function like `+[](GLFWwindow*, int width, int height){/*...*/}`

Diesel-Hadez commented on src/engine/window.hpp:
```diff
@@ -0,0 +1,42 @@
+   */
+  void update();
+
+  void *getNativeWindow();
```
Why do you return void* instead of GLFWwindow* ????????

Also, for what reason do you have this functionality? Seems kind of a leaky abstraction if a lot of other code requires a pointer to the original GLFWwindow* to make it work.

Diesel-Hadez commented on src/engine/window.hpp:
```diff
@@ -0,0 +1,42 @@
+  void *getNativeWindow();
+
+ private:
+  GLFWwindow *m_Window;
```
```suggestion
  GLFWwindow *m_Window {nullptr};
```
I think this style instead of doing it in the constructor has been available since C++11. Just a default value for it.

Diesel-Hadez commented on src/engine/window.hpp:
```diff
@@ -0,0 +1,42 @@
+  /*
+   * Functions
+   */
+  void update();
```
const correctness

Diesel-Hadez commented on src/engine/window.hpp:
```diff
@@ -0,0 +1,42 @@
+  /*
+   * Constructors/Operators
+   */
+  static std::expected<std::unique_ptr<Window>, EngineError> create(
```
Add [[nodiscard]]

Maybe also start using aliases since the types is long (e.g: `using WindowPtr=std::unique_ptr<Window>` so you can just have `std::expected<WindowPtr, EngineError>`)

Diesel-Hadez commented on src/engine/error.hpp:
```diff
@@ -0,0 +1,15 @@
+#ifndef ENGINEERROR_H
+#define ENGINEERROR_H
+
+enum class EngineError {
```
Rename this filename to `EngineError.hpp` so it's easy to telescope it, when you have multiple other error header files for other purposes (e.g: Another `error.hpp` for Playground Errors in the playground directory, another `error.hpp` in the shader directory etc.)

Diesel-Hadez commented on src/main.cpp:
```diff
@@ -0,0 +1,22 @@
+      .Title = std::string("Playground!"), .Width = 800, .Height = 600};
+  auto res = PlaygroundApp::create(spec);
+  if (!res.has_value()) {
+    std::cerr << "App creation error" << std::endl;
```
You can log the exact error by, for example, using [magic_enum](https://github.com/Neargye/magic_enum) (external library).

```cpp
std::cerr << std::format("App creation error. [error={}]", magic_enum::enum_name(res.error())) << std::endl;
```

Alternatively, if you don't want to use that and handle it manually, use a switch over the different possible values of the enum. Also, set up your LSP/Warnings so that every switch case on an enum has to cover all possible possibilities. 
```cpp
const auto errorName = [res](){
    switch (res.error())
    {
        case XYZ: return "XYZ";
        case ABC_1: [[fallthrough]];
        case ABC_2: return "ABC";
     }
    std::unreachable();
}();
std::cerr << std::format("App creation error. [error={}]", magic_enum::enum_name(errorName)) << std::endl;
```

Uh I was too lazy to wrap the cases in the switch in braces, but you should probably do that, depending on your linter settings.

Diesel-Hadez commented on src/main.cpp:
```diff
@@ -0,0 +1,22 @@
+  auto res = PlaygroundApp::create(spec);
+  if (!res.has_value()) {
+    std::cerr << "App creation error" << std::endl;
+    std::exit(1);
```
Maybe use `EXIT_FAILURE` and `EXIT_SUCCESS`? I usually prefer -1 for errors, up to you.

Diesel-Hadez commented on src/main.cpp:
```diff
@@ -0,0 +1,22 @@
+  auto app = std::move(*res);
+
+  app->registerProgram(std::make_unique<TriangleProgram>());
+  app->runProgram("Triangle");
```
You might want some type-safety here and have runProgram take in an enumeration of Programs. I can see why you might have did that though, in case you wanted to be able to create programs in the future that are created at runtime (and thus, named at runtime).

Diesel-Hadez commented on src/engine/application.hpp:
```diff
@@ -0,0 +1,30 @@
+#ifndef ENGINE_APPLICATION_H
+#define ENGINE_APPLICATION_H
```
If you want to use the Engine in separate projects in the future, you might want to make it a separate library, and make your program link to this "Engine".

Diesel-Hadez commented on src/playground_app.cpp:
```diff
@@ -0,0 +1,195 @@
+#include <backends/imgui_impl_glfw.h>
+#include <backends/imgui_impl_opengl3.h>
+
+constexpr uint8_t OPENGL_VERSION{41};
```
Wrap in anonymous namespace.

Diesel-Hadez commented on src/playground_app.hpp:
```diff
@@ -0,0 +1,47 @@
+  Program *SelectedProgram;
+  /// A flag to trigger reloading of program based on currently selected
+  /// program.
+  bool Run = false;
```
I prefer `{}` but that's a style thing. Also you might want to default your `SelectedProgram` to `nullptr` (If your LSP is not complaining about this yet, fix it).

Diesel-Hadez commented on src/playground_app.hpp:
```diff
@@ -0,0 +1,47 @@
+class PlaygroundApp : public Application {
+ public:
+  ~PlaygroundApp() override;
+  static std::expected<std::unique_ptr<PlaygroundApp>, EngineError> create(
```
Add `[[nodiscard]]`

Diesel-Hadez replied:
Also the same thing I said earlier about type aliases for shorter names.

Diesel-Hadez commented on src/playground_app.cpp:
```diff
@@ -0,0 +1,195 @@
+  // Loop
+  while (!glfwWindowShouldClose(window)) {
+    glfwPollEvents();
+    glClearColor(0.0f, 0.0f, 0.0f, 1.0f);
```
Magic values. Maybe set it to a BackgroundColour variable of some sort.

Diesel-Hadez commented on src/playground_app.cpp:
```diff
@@ -0,0 +1,195 @@
+}
+
+void PlaygroundApp::runProgram(const std::string& name) {
+  // Why do we need 2 lookups.
+  if (m_Programs.find(name) == m_Programs.end()) {
+    std::cerr << std::format("No program {} found.", name);
+    return;
+  }
+
+  // Cleanup old program
+  if (m_ActiveProgram != nullptr) {
+    m_ActiveProgram->cleanup();
+  }
+
+  m_ActiveProgram = m_Programs.at(name).get();
+  m_ActiveProgram->setup();
```
You don't need 2 lookups. You can just

```suggestion
  // Why do we need 2 lookups.
  if (auto program = m_Programs.find(name); 
          program == m_Programs.end()) {
    std::cerr << std::format("No program {} found.", name);
    return;
  }

  // Cleanup old program
  if (m_ActiveProgram != nullptr) {
    m_ActiveProgram->cleanup();
  }

  m_ActiveProgram = program;
  m_ActiveProgram->setup();
```

Actually program will go out of scope, you can fix that manually I guess.

Diesel-Hadez commented on src/playground_app.cpp:
```diff
@@ -0,0 +1,195 @@
+      ImGui::BeginChild("Programs", ImVec2(150, 0),
+                        ImGuiChildFlags_Borders | ImGuiChildFlags_ResizeX);
+
+      // Some c++17 arcane shit that I just discovered.
```
This is not that arcane, fairly common.

LemonTea387 replied:
I'm new okay.

Diesel-Hadez commented on src/playground_app.cpp:
```diff
@@ -0,0 +1,195 @@
+}
+
+void PlaygroundApp::showProgramSelector() {
+  ImGui::SetNextWindowSize(ImVec2(500, 440), ImGuiCond_FirstUseEver);
```
All the magic values like the window sizes for imgui in this file can be extracted to a variable, preferably const/constexpr.

Diesel-Hadez commented on src/playground_app.cpp:
```diff
@@ -0,0 +1,195 @@
+      ImGui::BeginGroup();
+      ImGui::BeginChild(
+          "Details",
+          ImVec2(0, -ImGui::GetFrameHeightWithSpacing()));  // Leave room for 1
+                                                            // line below us
```
I'd suggest not to put comments on the right of statements, and for comments to be on dedicated lines, for readability.

Diesel-Hadez commented on src/programs/program.hpp:
```diff
@@ -0,0 +1,27 @@
+  virtual void setup() = 0;
+  virtual void loop() = 0;
+  virtual void cleanup() = 0;
+  const std::string &getDescription() const { return m_Description; };
+  const std::string &getName() const { return m_Name; };
```
```suggestion
  [[nodiscard]] const std::string &getDescription() const { return m_Description; };
  [[nodiscard]] const std::string &getName() const { return m_Name; };
```

Diesel-Hadez commented on src/programs/triangle.cpp:
```diff
@@ -0,0 +1,50 @@
+
+  auto shader = Shader::create("triangle", "./shaders/triangle.vert.glsl",
+                               "./shaders/triangle.frag.glsl");
+  if (!shader.has_value()) {
+    return;
+  }
```
You can probably log the error (if it isn't logged by the shader itself, I forgot if it is) or propagate he error upwards and make this function return an expected<void,error> type. Callers would probably be confused if their setup quietly failed.

Diesel-Hadez commented on src/programs/triangle.hpp:
```diff
@@ -0,0 +1,24 @@
+
+#include <glad/gl.h>
+#include <GL/gl.h>
+#include "engine/renderer/shader.hpp"
```
Not needed. Just forward-declare.

Diesel-Hadez commented on src/programs/triangle.cpp:
```diff
@@ -0,0 +1,50 @@
+    return;
+  }
+  m_Shader = std::move(*shader);
+  m_Active = true;
```
If you want, you can move this in the header too. Stylistic thing though I guess.

Diesel-Hadez reviewed (commented):
Some light feedback.

LemonTea387 committed:
well it's updated

LemonTea387 committed:
std for consistency cstdint

LemonTea387 committed:
cleanup with fwd declrs and includes

LemonTea387 committed:
clarity on exit codes

LemonTea387 committed:
anonymous namespace for it being local

LemonTea387 committed:
rename to engine error for clarity

LemonTea387 committed:
use internal type alias for easy reading

LemonTea387 committed:
maybe I could make it generic use

LemonTea387 committed:
make them const and mark nodiscard

LemonTea387 committed:
setup now has error type

LemonTea387 committed:
use initializer

also a todo

LemonTea387 committed:
user level errors!

LemonTea387 committed:
enforce Warnings and fixing a bunch of them

LemonTea387 committed:
rmb to clean up this

LemonTea387 committed:
hey new example

LemonTea387 committed:
double the texture double the fun

LemonTea387 committed:
why call again?

LemonTea387 committed:
unflip textures

LemonTea387 committed:
remove some of the ported old examples

LemonTea387 committed:
make glad pulled externally.

LemonTea387 committed:
cleanup of cmakelist

use fetch content for everything now except imgui. Will figure out smth
with it later.

LemonTea387 committed:
somehow missed this

LemonTea387 committed:
init nullptr

LemonTea387 committed:
no discard getters

LemonTea387 committed:
remove this comment

LemonTea387 committed:
init

LemonTea387 committed:
extract magic values out

LemonTea387 committed:
some test readme changes

LemonTea387 committed:
update some goals
